### ISSUE TITLE
[ray] Build failure when installing from source

### ISSUE THREAD

**Issue by acxz** [ray-project/ray#8535]
### What is the problem?

I am trying to install ray from source. However, I receive the following error during the building process:

```
INFO: From Executing genrule //:python/ray/streaming/_streaming.pyx_cython_translation:
warning: python/ray/streaming/includes/transfer.pxi:57:50: Not all members given for struct 'CChannelCreationParameter'
warning: python/ray/streaming/includes/transfer.pxi:57:50: Not all members given for struct 'CChannelCreationParameter'
ERROR: /home/acxz/.cache/yay/python-ray/src/ray-ray-0.8.5/BUILD.bazel:371:1: Linking of rule '//:gcs_server' failed (Exit 1): gcc failed: error executing command 
  (cd /home/acxz/.cache/bazel/_bazel_acxz/e78727c45c5a88c44dd6ec8fadd46e6f/sandbox/linux-sandbox/1714/execroot/com_github_ray_project_ray && \
  exec env - \
    LD_LIBRARY_PATH=/home/acxz/cf_ws/devel/lib:/opt/ros/melodic/lib:/usr/lib/gazebo-11/plugins:/usr/lib/gazebo-11/plugins \
    PATH=/home/acxz/cf_ws/devel/bin:/opt/ros/melodic/bin:/home/acxz/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/opt/cuda/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl \
    PWD=/proc/self/cwd \
  /usr/bin/gcc @bazel-out/k8-opt/bin/gcs_server-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox gcc failed: error executing command 
  (cd /home/acxz/.cache/bazel/_bazel_acxz/e78727c45c5a88c44dd6ec8fadd46e6f/sandbox/linux-sandbox/1714/execroot/com_github_ray_project_ray && \
  exec env - \
    LD_LIBRARY_PATH=/home/acxz/cf_ws/devel/lib:/opt/ros/melodic/lib:/usr/lib/gazebo-11/plugins:/usr/lib/gazebo-11/plugins \
    PATH=/home/acxz/cf_ws/devel/bin:/opt/ros/melodic/bin:/home/acxz/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/opt/cuda/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl \
    PWD=/proc/self/cwd \
  /usr/bin/gcc @bazel-out/k8-opt/bin/gcs_server-2.params)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-opt/bin/_objs/gcs/subscription_executor.pic.o:subscription_executor.cc:function ray::gcs::SubscriptionExecutor<ray::ClientID, ray::rpc::HeartbeatTableData, ray::gcs::HeartbeatTable>::AsyncSubscribeAll(ray::ClientID const&, std::function<void (ray::ClientID const&, ray::rpc::HeartbeatTableData const&)> const&, std::function<void (ray::Status)> const&): error: undefined reference to 'ray::gcs::Log<ray::ClientID, ray::rpc::HeartbeatTableData>::Subscribe(ray::JobID const&, ray::ClientID const&, std::function<void (ray::gcs::RedisGcsClient*, ray::ClientID const&, std::vector<ray::rpc::HeartbeatTableData, std::allocator<ray::rpc::HeartbeatTableData> > const&)> const&, std::function<void (ray::gcs::RedisGcsClient*)> const&)'
bazel-out/k8-opt/bin/_objs/gcs/subscription_executor.pic.o:subscription_executor.cc:function ray::gcs::SubscriptionExecutor<ray::ClientID, ray::rpc::HeartbeatBatchTableData, ray::gcs::HeartbeatBatchTable>::AsyncSubscribeAll(ray::ClientID const&, std::function<void (ray::ClientID const&, ray::rpc::HeartbeatBatchTableData const&)> const&, std::function<void (ray::Status)> const&): error: undefined reference to 'ray::gcs::Log<ray::ClientID, ray::rpc::HeartbeatBatchTableData>::Subscribe(ray::JobID const&, ray::ClientID const&, std::function<void (ray::gcs::RedisGcsClient*, ray::ClientID const&, std::vector<ray::rpc::HeartbeatBatchTableData, std::allocator<ray::rpc::HeartbeatBatchTableData> > const&)> const&, std::function<void (ray::gcs::RedisGcsClient*)> const&)'
bazel-out/k8-opt/bin/_objs/gcs/subscription_executor.pic.o:subscription_executor.cc:function ray::gcs::SubscriptionExecutor<ray::WorkerID, ray::rpc::WorkerFailureData, ray::gcs::WorkerFailureTable>::AsyncSubscribeAll(ray::ClientID const&, std::function<void (ray::WorkerID const&, ray::rpc::WorkerFailureData const&)> const&, std::function<void (ray::Status)> const&): error: undefined reference to 'ray::gcs::Log<ray::WorkerID, ray::rpc::WorkerFailureData>::Subscribe(ray::JobID const&, ray::ClientID const&, std::function<void (ray::gcs::RedisGcsClient*, ray::WorkerID const&, std::vector<ray::rpc::WorkerFailureData, std::allocator<ray::rpc::WorkerFailureData> > const&)> const&, std::function<void (ray::gcs::RedisGcsClient*)> const&)'
collect2: error: ld returned 1 exit status
Target //:ray_pkg failed to build
INFO: Elapsed time: 1150.801s, Critical Path: 83.46s
INFO: 1710 processes: 1710 linux-sandbox.
FAILED: Build did NOT complete successfully
Traceback (most recent call last):
  File "setup.py", line 202, in <module>
    setup(
  File "/usr/lib/python3.8/site-packages/setuptools/__init__.py", line 144, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/usr/lib/python3.8/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/usr/lib/python3.8/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 121, in run
    subprocess.check_call(command)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['../build.sh', '-p', '/usr/bin/python']' returned non-zero exit status 1.

```

*Ray version and other system information (Python version, TensorFlow version, OS):*
ray: 0.8.5
python: 3.8.3
gcc: 10.1.0
bazel: 3.1.0
os: Arch Linux

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

The above does not apply since this is a build issue.

**Comment by rkooo567** [comment]
@simon-mo

**Comment by simon-mo** [comment]
Seems like GCS build issue. @ffbin can you take a look or find someone might touch this code recently? Thanks

The error is undefined symbol, reformatted the log below.

```
bazel-out/k8-opt/bin/_objs/gcs/subscription_executor.pic.o:subscription_executor.cc:function ray::gcs::SubscriptionExecutor<ray::ClientID, ray::rpc::HeartbeatTableData, ray::gcs::HeartbeatTable>::AsyncSubscribeAll(ray::ClientID const&, std::function<void (ray::ClientID const&, ray::rpc::HeartbeatTableData const&)> const&, std::function<void (ray::Status)> const&): 
error: undefined reference to 'ray::gcs::Log<ray::ClientID, ray::rpc::HeartbeatTableData>::Subscribe(ray::JobID const&, ray::ClientID const&, std::function<void (ray::gcs::RedisGcsClient*, ray::ClientID const&, std::vector<ray::rpc::HeartbeatTableData, std::allocator<ray::rpc::HeartbeatTableData> > const&)> const&, std::function<void (ray::gcs::RedisGcsClient*)> const&)'

bazel-out/k8-opt/bin/_objs/gcs/subscription_executor.pic.o:subscription_executor.cc:function ray::gcs::SubscriptionExecutor<ray::ClientID, ray::rpc::HeartbeatBatchTableData, ray::gcs::HeartbeatBatchTable>::AsyncSubscribeAll(ray::ClientID const&, std::function<void (ray::ClientID const&, ray::rpc::HeartbeatBatchTableData const&)> const&, std::function<void (ray::Status)> const&): 
error: undefined reference to 'ray::gcs::Log<ray::ClientID, ray::rpc::HeartbeatBatchTableData>::Subscribe(ray::JobID const&, ray::ClientID const&, std::function<void (ray::gcs::RedisGcsClient*, ray::ClientID const&, std::vector<ray::rpc::HeartbeatBatchTableData, std::allocator<ray::rpc::HeartbeatBatchTableData> > const&)> const&, std::function<void (ray::gcs::RedisGcsClient*)> const&)'

bazel-out/k8-opt/bin/_objs/gcs/subscription_executor.pic.o:subscription_executor.cc:function ray::gcs::SubscriptionExecutor<ray::WorkerID, ray::rpc::WorkerFailureData, ray::gcs::WorkerFailureTable>::AsyncSubscribeAll(ray::ClientID const&, std::function<void (ray::WorkerID const&, ray::rpc::WorkerFailureData const&)> const&, std::function<void (ray::Status)> const&): 
error: undefined reference to 'ray::gcs::Log<ray::WorkerID, ray::rpc::WorkerFailureData>::Subscribe(ray::JobID const&, ray::ClientID const&, std::function<void (ray::gcs::RedisGcsClient*, ray::WorkerID const&, std::vector<ray::rpc::WorkerFailureData, std::allocator<ray::rpc::WorkerFailureData> > const&)> const&, std::function<void (ray::gcs::RedisGcsClient*)> const&)'
```

**Comment by acxz** [comment]
Spent a little more time into this today and I think I have a solution (#8690), not too sure tho, since my build fails but now it fails somewhere else that I believe is unrelated. If someone could test my changes that would be great.

**Comment by simon-mo** [comment]
I just merged #8690 since the test failure seems unrelated.

**Comment by acxz** [comment]
Thx so much!
